### ISSUE TITLE
Possible to specify output directory and module path structure?

### ISSUE THREAD

**Issue by BenRKarl** [googleapis/gapic-generator-python#297]
Currently the microgenerator outputs code in the below directory structure when I provide the following CLI flag: `--python_gapic_opt="lazy-import,python-gapic-name=google_ads"`

```
google/
    ads/
        google_ads/
        google_ads_v2/
```

which results in the import paths looking like: `google.ads.google_ads_v2 ...`

Is it possible to just generate the code into a single directory with the given version number? So to be specific it would look like:

```
v2/
```

The reason is that we place the generated code into an existing structure that looks like this:

```
google/
    ads/
        google_ads/
            v1/
            v2/
```

And all the existing import paths in the generated code should look like:

`google.ads.google_ads.v2 ...`

I'm wondering if it's possible to specify this structure as an option, or if it requires significant code changes in the generator.

**Comment by software-dov** [comment]
I believe this is possible but tricky for technical reasons dealing with the current implementation and template layout. Part of the reason for the existing layout is to provide a user-transparent upgrade path with a versioned fallback: user code can do `from google.ads import google_ads`, and then if a new version comes out they don't need to change their import paths. Functionally, it's just an alias to the versioned module, which they can also name explicitly. 
I kind of want to push back and say "no, everybody gets the same module layout,"  The transparent live-at-head nature will hopefully make it easier for users to upgrade in the future. How much of a dealbreaker is the directory layout?

**Comment by BenRKarl** [comment]
Thanks for the info! I sort of expected as much. The design of our library sort of makes the user-transparent upgrade path a bit moot, as users don't explicitly import versioned modules - retrieval of versioned protobuf message modules and classes is done through the client itself [here](https://github.com/googleads/google-ads-python/blob/master/google/ads/google_ads/client.py#L209) and [here](https://github.com/googleads/google-ads-python/blob/master/google/ads/google_ads/client.py#L162). So that means it's not a huge deal if the underlying dir structure changes because we can refactor the client and keep the same user-facing surface. The tricky part is that when we move the generated files into a new folder structure then the import paths in the generated code can be inaccurate. I think we can deal with a different layout, I'll talk with my team and let you know if any other problems come up around this.

**Comment by busunkim96** [comment]
@BenRKarl Cloud uses the directory structure produced by the generator. [see example](https://github.com/googleapis/python-asset/tree/master/google/cloud)

We prefer it as it allows folks to get the latest stable version without having to edit their code.

```python
from google.cloud import asset  # defaults to v1 now, previously defaulted to v1beta1
```

**Comment by software-dov** [comment]
This may turn out to be a bad idea, but I think we may be able to have our cake and eat it too.
Desirables:
* Multiple versions of the same module, differentiated via `v1`, `v2`, etc.
* A 'default' module that points to the most recent version as an alias/symlink.

The tricky bit is that users may deviate from the suggested interface, in which case the actual module layout becomes visible and changes become breaking. I'll whip up an alternatives matrix.

**Comment by busunkim96** [comment]
Cloud samples generally have imports of the following format. Anyone who imports using something other than the default `from google.cloud import asset` will be broken. 

```py
from google.cloud import asset_v1 as asset
```

TLDR it isn't uncommon for someone to be importing from a specific version instead of the default. 

If there can only be one directory structure it looks like it will have to break either Ads or Cloud.

**Comment by software-dov** [comment]
Which is exactly what neither party wants :P Which means that we're left with a generator option and sweeping the nastiness into the generator instead of onto our users or ancillary code.

**Comment by software-dov** [comment]
Handled by a combination of new features (#341) and custom templates. Closed as fixed.
